jenkins-master/jobs/scripts/workspace/runtests.py

```diff
@@ -5,85 +5,84 @@
+    'TREEHERDER_URL', 'TREEHERDER_CLIENT_ID', 'TREEHERDER_SECRET',
+]
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--repository',
+                        required=True,
+                        help='The repository name the build was created from.')
+    parser.add_argument('--test-type',
+                        required=True,
+                        choices=config['test_types'].keys(),
+                        help='The type of tests to execute')
+    parser.add_argument('--installer-url',
+                        required=True,
+                        help='The URL of the build installer.')
 
-class Runner(object):
+    update_group = parser.add_argument_group('update', 'Update test specific options')
+    update_group.add_argument('--update-channel',
+                              action=JenkinsDefaultValueAction,
+                              help='The update channel to use for the update test')
+    update_group.add_argument('--update-target-build-id',
+                              action=JenkinsDefaultValueAction,
+                              help='The expected BUILDID of the updated build')
+    update_group.add_argument('--update-target-version',
+                              action=JenkinsDefaultValueAction,
+                              help='The expected version of the updated build')
+    return parser.parse_args()
 
-    def run_tests(self, args):
-        settings = config['test_types'][args.type]
 
-        # In case the log folder does not exist yet we have to create it because
-        # otherwise Marionette will fail for e.g. gecko.log (see bug 1211666)
-        if settings['logs'].get('gecko.log'):
-            try:
-                os.makedirs(os.path.dirname(settings['logs']['gecko.log']))
-            except OSError:
-                print('Failed to create log folder for {}'.format(settings['logs']['gecko.log']))
+def main():
+    args = parse_args()
 
-        print('Downloading the installer: {}'.format(args.installer_url))
-        scraper = FactoryScraper('direct',
-                                 url=args.installer_url,
-                                 retry_attempts=5,
-                                 retry_delay=30,
-                                 )
-        installer_path = scraper.download()
+    try:
+        settings = config['test_types'][args.test_type]
+        retval = 0
 
-        command = [
-            'firefox-ui-update' if args.type == 'update' else 'firefox-ui-tests',
-            '--installer', installer_path,
-            '--workspace', os.getcwd(),
-            '--log-tbpl', settings['logs']['tbpl.log'],
-        ]
+        command = [os.path.join('mozharness', 'scripts', settings['harness_script']),
+                   '--cfg', settings['harness_config'],
+                   '--firefox-ui-branch', args.repository,
+                   '--installer-url', args.installer_url,
+                   ]
 
-        if args.type == 'update':
```

Comment:
I pushed a new commit which refactors runtests.py and introduces runner classes.